include/wil/windowing.h

```diff
@@ -0,0 +1,164 @@
+#ifdef __cpp_if_constexpr
+            using result_t = decltype((*pCallback)(hwnd));
+            if constexpr (std::is_void_v<result_t>)
+            {
+                (*pCallback)(hwnd);
+                return TRUE;
+            }
+            else if constexpr (std::is_same_v<result_t, HRESULT>)
+            {
+                // NB: this works for both HRESULT and NTSTATUS as both S_OK and ERROR_SUCCESS are 0
+                return (S_OK == (*pCallback)(hwnd)) ? TRUE : FALSE;
+            }
+            else if constexpr (std::is_same_v<result_t, bool>)
+            {
+                return (*pCallback)(hwnd) ? TRUE : FALSE;
+            }
+            else
+            {
+                static_assert(details::always_false<TCallback>::value, "Callback must return void, bool, or HRESULT");
+            }
+#else
+            return (*pCallback)(hwnd);
+#endif
+        };
+        enumApi(enumproc, reinterpret_cast<LPARAM>(&callback));
+    }
+
+#ifdef WIL_ENABLE_EXCEPTIONS
+    template <typename TEnumApi, typename TCallback>
+    void CallCallback(TEnumApi&& enumApi, TCallback&& callback)
+    {
+        struct
+        {
+            std::exception_ptr exception;
+            TCallback* pCallback;
+        } callbackData = {nullptr, &callback};
+        auto enumproc = [](HWND hwnd, LPARAM lParam) -> BOOL {
+            auto pCallbackData = reinterpret_cast<decltype(&callbackData)>(lParam);
+            try
+            {
+                auto pCallback = pCallbackData->pCallback;
+#ifdef __cpp_if_constexpr
+                using result_t = decltype((*pCallback)(hwnd));
+                if constexpr (std::is_void_v<result_t>)
+                {
+                    (*pCallback)(hwnd);
+                    return TRUE;
+                }
+                else if constexpr (std::is_same_v<result_t, HRESULT>)
+                {
+                    // NB: this works for both HRESULT and NTSTATUS as both S_OK and ERROR_SUCCESS are 0
+                    return (S_OK == (*pCallback)(hwnd)) ? TRUE : FALSE;
+                }
+                else if constexpr (std::is_same_v<result_t, bool>)
+                {
+                    return (*pCallback)(hwnd) ? TRUE : FALSE;
+                }
+                else
+                {
+                    static_assert(details::always_false<TCallback>::value, "Callback must return void, bool, or HRESULT");
+                }
+#else
+                return (*pCallback)(hwnd);
+#endif
+            }
+            catch (...)
+            {
+                pCallbackData->exception = std::current_exception();
+                return FALSE;
+            }
+        };
+        enumApi(enumproc, reinterpret_cast<LPARAM>(&callbackData));
+        if (callbackData.exception)
+        {
+            std::rethrow_exception(callbackData.exception);
+        }
+    }
+#endif
+} // namespace details
+
+template <typename TCallback>
+auto for_each_window_nothrow(TCallback&& callback) noexcept

```

Comment:
All these have an `auto` return type, but don't actually `return` anything. If intentional, explicitly mark as `void`; gives off the impression that something was forgotten

Reply:
changed to void